### PR TITLE
Fix assertion "key not already in map".

### PR DESCRIPTION
**Pull request by schweitzpgi**
Fix verification of fir.if bug.
Some progress on the symbol not mapped cases.
Add some TODOs.
Add forall-2.f90 test file.

### CODE REVIEW

File: flang/lib/Lower/ConvertExpr.cpp [hunk 1]
```diff
@@ -2416,10 +2416,17 @@ class ArrayExprLowering {
   };
 
   using ExtValue = fir::ExtendedValue;
-  using IterSpace = const IterationSpace &;      // active iteration space
-  using CC = std::function<ExtValue(IterSpace)>; // current continuation
-  using PC =
-      std::function<IterationSpace(IterSpace)>; // projection continuation
+  /// Active iteration space.
+  using IterSpace = const IterationSpace &;
+  /// Current continuation. Function that we generate IR for a single iteration
```
Comment (vdonaldson) [inline flang/lib/Lower/ConvertExpr.cpp:2421]: we -> will ?

File: flang/lib/Lower/ConvertExpr.cpp [hunk 2]
```diff
@@ -2416,10 +2416,17 @@ class ArrayExprLowering {
...
+  /// of the pending iterative loop structure.
+  using CC = std::function<ExtValue(IterSpace)>;
+  /// Projection continuation. Function that will project one iteration space
+  /// into another.
+  using PC = std::function<IterationSpace(IterSpace)>;
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:2426]: nit: `IterSpace(IterSpace)` or `IterationSpace(IterationSpace)` for consistency.

File: flang/lib/Lower/ConvertExpr.cpp [hunk 3]
```diff
@@ -3203,6 +3212,51 @@ class ArrayExprLowering {
     return {indices, loops[0]};
   }
 
+  llvm::SmallVector<mlir::Value> genImplicitLoopBounds(
+      const Fortran::evaluate::Expr<Fortran::evaluate::SomeType> &e) {
+    struct Filter : public Fortran::evaluate::AnyTraverse<
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:3217]: What is guaranteeing that the `ArrayRef` in `e` that gets visited is the one that matters here ?

File: flang/lib/Lower/ConvertExpr.cpp [hunk 4]
```diff
@@ -4394,11 +4450,54 @@ class ArrayExprLowering {
...
+      return {};
+    return getFrontEndShape(x);
+  }
+  template <typename A>
+  llvm::SmallVector<mlir::Value> getFrontEndShape(const A &x) {
+    if (auto optShape = Fortran::evaluate::GetShape(x)) {
+      llvm::SmallVector<mlir::Value> result;
+      convertFEShape(*optShape, result);
+      if (!result.empty())
+        return result;
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:4492]: What is the purpose of this `if (!result.empty())` ? If result is empty, it looks like an empty vector is returned anyway.

File: flang/lib/Lower/ConvertExpr.cpp [hunk 5]
```diff
@@ -4394,11 +4450,54 @@ class ArrayExprLowering {
         }(),
         x);
   }
+  static mlir::Type unwrapBoxEleTy(mlir::Type ty) {
+    if (auto boxTy = ty.dyn_cast<fir::BoxType>()) {
+      ty = boxTy.getEleTy();
+      if (auto refTy = fir::dyn_cast_ptrEleTy(ty))
+        ty = refTy;
+    }
+    return ty;
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:4459]: Wouldn't `fir::dyn_cast_ptrOrBoxEleTy()` yield the same result on the intended use cases of `unwrapBoxEleTy` (fir.box or non memory `ty`) ?

File: flang/lib/Lower/ConvertExpr.cpp [hunk 6]
```diff
@@ -4412,30 +4511,64 @@ class ArrayExprLowering {
...
+                                  : builder.createIntegerConstant(loc, idxTy,
+                                                                  1);
+                          auto step = builder.createConvert(
+                              loc, idxTy,
+                              fir::getBase(asScalar(triple.stride())));
+                          auto diff = builder.create<mlir::SubIOp>(loc, up, lo);
+                          auto sum =
+                              builder.create<mlir::AddIOp>(loc, diff, step);
+                          mlir::Value count =
+                              builder.create<mlir::SignedDivIOp>(loc, sum,
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:4564]: This triplet extent computation pattern will definitely deserve to be centralized in a helper at some point. FYI, I think it does not cover the legal case of `i:j:k` where `j-i` and k sign differ, in which case the computed extent must be zero.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 3 above)
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:3217]: You're right. This is a fake "analysis" here. I think I'll have to write one from scratch to get what I really need.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 2 above)
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:2426]: They are different types however.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 4 above)
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:4492]: Thanks. It was left over from some other experimenting.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 6 above)
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:4564]: Agreed. I'll put that on my list of TODOs.

**Review by vdonaldson — APPROVED** [review]

**Pull request merged**